apis/python/src/tiledbsoma/util_tiledb.py

```diff
@@ -57,47 +57,59 @@ def to_tiledb_supported_array_type(x: T) -> T:
-        )
-
-    # If a Pandas categorical, use the type of the underlying category.
-    # If the array contains NaN/NA, and the primitive is unable to represent
-    # a reasonable facsimile, i.e. not string or float, raise.
-    if pd.api.types.is_categorical_dtype(x.dtype):
-        assert isinstance(x, pd.Series)
-        categories = x.cat.categories
-        cat_dtype = categories.dtype
-        if cat_dtype.kind in ["f", "u", "i"]:
-            if x.hasnans and cat_dtype.kind == "i":
-                raise ValueError(
-                    "Categorical array contains NaN -- unable to convert to TileDB array."
-                )
-
-            return x.astype(to_tiledb_supported_dtype(cat_dtype))
-
-        # Into the weirdness. See if Pandas can help with edge cases.
-        inferred = pd.api.types.infer_dtype(categories)
-        if inferred == "boolean":
-            if x.hasnans:
-                raise ValueError(
-                    "Categorical array contains NaN -- unable to convert to TileDB array."
-                )
-            return x.astype(bool)
-
-        if inferred == "string":
-            return x.astype(str)
-
-        if inferred == "bytes":
-            if x.hasnans:
-                raise ValueError(
-                    "Categorical array contains NaN -- unable to convert to TileDB array."
-                )
-            return x.astype(bytes)
-
-        return x.astype("O")
+    return _to_supported_base(x)
+
+
+@to_tiledb_supported_array_type.register
+def _to_supported_dataframe(x: pd.DataFrame) -> pd.DataFrame:
+    return pd.DataFrame.from_dict(
+        {k: to_tiledb_supported_array_type(v) for k, v in x.items()}
+    )
+
+
+@to_tiledb_supported_array_type.register
+def _to_supported_series(x: pd.Series) -> pd.Series:
+    if not pd.api.types.is_categorical_dtype(x.dtype):
+        return _to_supported_base(x)
+
+    categories = x.cat.categories
+    cat_dtype = categories.dtype
+    if cat_dtype.kind in ["f", "u", "i"]:
+        if x.hasnans and cat_dtype.kind == "i":
+            raise ValueError(
+                "Categorical array contains NaN -- unable to convert to TileDB array."
+            )
+
+        return x.astype(to_tiledb_supported_dtype(cat_dtype))
+
+    # Into the weirdness. See if Pandas can help with edge cases.
+    inferred = pd.api.types.infer_dtype(categories)
+    if inferred == "boolean":
+        if x.hasnans:
+            raise ValueError(
+                "Categorical array contains NaN -- unable to convert to TileDB array."
+            )
+        return x.astype(bool)
+
+    if inferred == "string":
+        return x.astype(str)
+
+    if inferred == "bytes":
+        if x.hasnans:
+            raise ValueError(
+                "Categorical array contains NaN -- unable to convert to TileDB array."
+            )
+        return x.astype(bytes)
+
+    return x.astype("O")
+
+
+_MT = TypeVar("_MT", np.ndarray, pd.Series, sp.spmatrix)  # type: ignore[type-arg]
```

Comment:
leftover from earlier work I was doing; I think this is the right thing to do:
```suggestion
_MT = TypeVar("_MT", np.ndarray, sp.spmatrix)  # type: ignore[type-arg]
```
Also the `Any -> Any` above might not be The Right Thing but it’s fine for now since it’s an internal utility function.